src/sage/schemes/elliptic_curves/ell_finite_field.py

```diff
@@ -2615,3 +2739,150 @@ def special_supersingular_curve(F, *, endomorphism=False):
     endo._degree = ZZ(q)
     endo.trace.set_cache(ZZ.zero())
     return E, endo
+
+def _EllipticCurve_with_order_helper(m, D):
+    r"""
+    See documentation of :func:`EllipticCurve_with_order`.
+
+    EXAMPLES::
+
+        sage: from sage.schemes.elliptic_curves.ell_finite_field import EllipticCurve_with_order
+        sage: E = EllipticCurve_with_order(1234); E  # random
+        Elliptic Curve defined by y^2 = x^3 + 1142*x + 1209 over Finite Field of size 1237
+        sage: E.order() == 1234
+        True
+    """
+    from sage.arith.misc import is_prime_power, factor
+    from sage.quadratic_forms.binary_qf import BinaryQF
+    from sage.structure.factorization import Factorization
+    from sage.schemes.elliptic_curves.cm import hilbert_class_polynomial
+
+    def helper(m, m4_fac, D):
+        all_sol = BinaryQF(1, 0, -D).solve_integer(m4_fac, flag=3)
+        for t, _ in all_sol:
+            yield m + 1 - t
+            yield m + 1 + t
+
+    if isinstance(m, Factorization):
+        m4_fac = m * factor(4)
+        m_val = m.value()
+    else:
+        m4_fac = factor(m * 4)
+        m_val = m
+
+    if D is None:
+        Ds = (D for D in range(-4 * m_val, 0) if D % 4 in [0, 1])
+    else:
+        assert D < 0 and D % 4 in [0, 1]
+        Ds = [D]
+
+    seen = set()
+    for D in Ds:
+        for q in helper(m_val, m4_fac, D):
+            if not is_prime_power(q):
+                continue
+
+            H = hilbert_class_polynomial(D)
+            K = GF(q)
+            roots = H.roots(ring=K)
+            for j0, _ in roots:
+                E = EllipticCurve(j=j0)
+                for Et in E.twists():
+                    if any(Et.is_isomorphic(E) for E in seen):
+                        continue
+                    try:
+                        # This tests whether the curve has given order
+                        Et.set_order(m_val)
+                        seen.add(Et)
+                        yield Et
+                    except ValueError:
+                        pass
+
+def EllipticCurve_with_order(m, *, D=None, all=False, iter=False):
+    r"""
+    Return an elliptic curve over a finite field with the given order. The curves are computed using
+    the Complex Multiplication (CM) method.
+
+    A `:sage:`~sage.structure.factorization.Factorization` can be passed for ``m``, in which case
+    the algorithm is more efficient.
+
+    If ``D`` is specified, it is used as the discriminant.
+
+    If ``all`` is set, a list of all elliptic curves over finite fields with the given order is
+    returned.
+
+    If ``iter`` is set, an iterator of all elliptic curves over finite fields with the given order
+    is returned. This cannot be set with ``all``.
```

Comment:
Always returning an iterator actually covers all three use cases using `next()` and `list()`, so I'm not convinced we need two keyword arguments just for this. (Iterators are not a very common sight in the Sage library, but I think they should be!)